### ISSUE TITLE
Strip trailing dot in FQDN during TLS validation

### ISSUE THREAD

**Issue by JJGadgets** [pgbackrest/pgbackrest#2442]
## Problem Description

I'm using pgBackRest as part of CrunchyData Postgres Operator, and due to Kubernetes ndots, I would like to add a trailing dot at the end of my S3 endpoint FQDNs to signify that the hostname in question is already fully qualified and thus doesn't need to recurse search domains.

However, this breaks TLS SNI validation as the hostname sent during validation has the trailing dot which doesn't match the certificate's SANs which doesn't have the trailing dot.

## Proposed Solution

I would like the trailing dot to be removed from the hostname sent during TLS SNI validation so the validation passes without the DNS query recursing search domains.

**Comment by dwsteele** [comment]
This seems like it could cause all sorts of breakage if applied across all our hosts/endpoints. We could add an option or options, but then there would be the question of which endpoints to apply the logic to.

We're pretty efficient with our lookups so I just don't see this leading to significant gains.

**Comment by JJGadgets** [comment]
> We're pretty efficient with our lookups so I just don't see this leading to significant gains.

Does pgBackRest cache DNS records across WAL pushes according to the records' TTL?

**Comment by dwsteele** [comment]
> Does pgBackRest cache DNS records across WAL pushes according to the records' TTL?

Unfortunately no, since `getaddrinfo()` does not provide this information. What we do is reuse connections that we already have. This allows backups (or whatever) to complete with relatively few connections being made.

Our TLS implementation is heavily based on Postgres' TLS implementation and to my knowledge Postgres does not do what you are proposing (it's possible this changed since I adapted the code, but I doubt it). If you can find a prominent open source project that is doing TLS validation this way we'd be happy to have a look. Or, if you can convince Postgres to make this change we'd likely follow suit.

As it is I think the risk is too high for the potential gain.

**Comment by dwsteele** [comment]
One more thing: our host name matching is based on RFC-2818 which according to my reading does not allow what you are proposing.

**Comment by dwsteele** [comment]
I did play around with this a bit. Disabling the RFC-2818 validation got past the first hurdle, but then I saw other issues -- for instance S3 will not accept host names in an HTTP header with a trailing dot. I'm guessing GCS and Azure are similar.

So, we'd really need to be careful about where to remove the dot (*almost* everywhere as far as I can see) and where to leave it in (*probably* just DNS resolution). This turns out to be harder than it looks. Probably the way to do it is to propagate HttpUrl throughout the codebase rather than pass hostname strings around, which has some advantages but definitely falls into the category of a big change.

I still believe the risks outweigh the benefits. Probably setting `ndots` appropriately in resolve.conf is easier overall.

We may revisit this in the future but for now I'm closing.
